src/lock.rs

```diff
@@ -0,0 +1,174 @@
+    package::{Package, PackageName},
+};
+
+pub const LOCKFILE: &str = "Proto.lock";
+
+/// Captures immutable metadata about a given package
+///
+/// It is used to ensure that future installations will use the exact same dependencies.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct LockedPackage {
+    pub name: PackageName,
+    pub checksum: String,
+    pub repository: String,
+    pub version: Version,
+    pub dependencies: Vec<PackageName>,
+}
+
+impl LockedPackage {
+    /// Captures the source, version and checksum of a Package for use in reproducible installs
+    pub fn lock(package: &Package, repository: String) -> Self {
+        let checksum = digest::digest(&digest::SHA256, &package.tgz);
+
+        Self {
+            name: package.name().to_owned(),
+            repository,
+            checksum: hex::encode(checksum),
+            version: package.version().to_owned(),
+            dependencies: package
+                .manifest
+                .dependencies
+                .iter()
+                .cloned()
+                .map(|d| d.package)
+                .collect(),
+        }
+    }
+
+    /// Validates if another LockedPackage matches this one
+    pub fn validate(&self, other: &Self) -> eyre::Result<()> {
+        ensure!(
+            self.name == other.name,
+            "Package name mismatch - expected {}, actual {}",
+            self.name,
+            other.name
+        );
+        ensure!(
+            self.checksum == other.checksum,
+            "Checksum mismatch - expected {}, actual {}",
+            self.checksum,
+            other.checksum
+        );
+        ensure!(
+            self.repository == other.repository,
+            "Repository mismatch - expected {}, actual {}",
+            self.repository,
+            other.repository
+        );
+        ensure!(
+            self.version == other.version,
+            "Version mismatch - expected {}, actual {}",
+            self.version,
+            other.version
+        );
+        ensure!(
+            self.dependencies == other.dependencies,
+            "Dependencies mismatch"
+        );
+        Ok(())
+    }
+
+    /// Constructs a Dependency instance with matching metadata
+    pub fn as_dependency(&self) -> Dependency {
+        Dependency {
+            package: self.name.clone(),
+            manifest: DependencyManifest {
+                version: VersionReq {
+                    comparators: vec![semver::Comparator {
+                        op: semver::Op::Exact,
+                        major: self.version.major,
+                        minor: Some(self.version.minor),
+                        patch: Some(self.version.patch),
+                        pre: self.version.pre.clone(),
+                    }],
+                },
+                repository: self.repository.clone(),
+            },
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct RawLockfile {
```

Comment:
I would be highly tempted to put a hash of the manifest into the lockfile, as Poetry does. Why? It means you can trivially detect whether the lockfile is out of date with the dependencies.

Reply:
Hmmm that's not necessarily true though, at least the way we're using the lockfile. Even if you add or remove some dependencies, it doesn't entirely invalidate the lockfile. The unchanged dependencies can be reused as they are. Same goes if you change the version requirement and they still resolve to the same versions.

I think we're copying the model Cargo uses, unless I've misunderstood something (quite possible).

Reply:
after further thought, i disagree with this suggestion now beacuse it unnecessarily generates git merge conflicts. However, the thing I was pointing out was not relating to entirely invalidating - rather that it makes it easier to check whether the lockfile is out of date and therefore in need of recomputing. It's not about it being partially valid per se.

Reply:
Ack, thanks for the clarification. That makes sense.